### ISSUE TITLE
Flask app without the development server in a sensor

### ISSUE THREAD

**Issue by NikosVlagoidis** [StackStorm/st2#4112]
Hello, I would like to know how to implement a Flask app inside a sensor with a WSGI server like gunicorn or gevent.WSGI Server. I tried both but it looks like there are some GIL issues and the server is blocked from returning a response. Any ideas?

**Comment by armab** [comment]
Would be best to post such question on https://forum.stackstorm.com/, instead of github issues which are for bugs/features/changes.

**Comment by NikosVlagoidis** [comment]
My use case is that I want to extend the st2 API with some custom functionality using a Flask App. For example, file uploads to the st2 server and then triggering some actions with this file. I don't want to use the development server that Flask provides. Creating a gunicorn and gevent WSGI server from inside a function leads to a bizarre result where the requests hang and I am guessing it is due to the GIL.

When I used the eventlet WSGI sever the problem solved and the requests went through normally.  My conclusion was that there is some conflict with the event loop that the frameworks are using but I don't know the subject in depth.

**Comment by NikosVlagoidis** [comment]
@armab, in my opinion, this is not just a question. This could be documented if someone else wants to use a Flask App to extend the API of st2. I have also seen packs using the dev server of Flask and I am not sure if that is the best practice.

**Comment by NikosVlagoidis** [comment]
@LindsayHill I can see your point. My approach is to add a location on the nginx and proxy the request to the Flask app.


Here is some code + logs.
 
# Using gevent

Request hangs no log output:

```python 
from flask import request, Flask
from gevent.pywsgi import WSGIServer

from st2reactor.sensor.base import Sensor


class EchoFlaskSensor(Sensor):
    def __init__(self, sensor_service, config):
        super(EchoFlaskSensor, self).__init__(
            sensor_service=sensor_service,
            config=config
        )

        self._host = '127.0.0.1'
        self._port = 5000
        self._path = '/echo'

        self._log = self._sensor_service.get_logger(__name__)
        self._app = Flask(__name__)

    def setup(self):
        pass

    def run(self):
        @self._app.route(self._path, methods=['POST'])
        def echo():

            return request.data

        self._log.info('Listening for payload on http://{}:{}{}'.format(
            self._host, self._port, self._path))
        # self._app.run(host=self._host, port=self._port, threaded=True)
        server = WSGIServer((self._host, self._port), self._app)
        server.serve_forever()

    def cleanup(self):
        pass

    def add_trigger(self, trigger):
        # This method is called when trigger is created
        pass

    def update_trigger(self, trigger):
        # This method is called when trigger is updated
        pass

    def remove_trigger(self, trigger):
        # This method is called when trigger is deleted
        pass
```
```bash
2018-05-05 07:28:00,182 INFO [-] Running sensor initialization code
2018-05-05 07:28:00,182 INFO [-] Running sensor in passive mode
2018-05-05 07:28:00,182 INFO [-] Listening for payload on http://127.0.0.1:5000/echo
```

```bash
root@a110dea09007:/ curl localhost:5000

```


# Using Gunicorn

```python
from flask import request, Flask
from gevent.pywsgi import WSGIServer
from gunicorn.app.base import BaseApplication

from st2reactor.sensor.base import Sensor


class Server(BaseApplication):

    def __init__(self, app):
        self.app = app
        super(Server, self).__init__()

    def load_config(self):
        return {"bind": "0.0.0.0:5000",
                "preload_app": False,
                "workers": 4,
                "worker_class": "sync",
                "max_requests": 10,
                "max_requests_jitter": 10}

    def load(self):
        return self.app


class EchoFlaskSensor(Sensor):
    def __init__(self, sensor_service, config):
        super(EchoFlaskSensor, self).__init__(
            sensor_service=sensor_service,
            config=config
        )

        self._host = '127.0.0.1'
        self._port = 5000
        self._path = '/echo'

        self._log = self._sensor_service.get_logger(__name__)
        self._app = Flask(__name__)

    def setup(self):
        pass

    def run(self):
        @self._app.route(self._path, methods=['POST'])
        def echo():
            return request.data

        self._log.info('Listening for payload on http://{}:{}{}'.format(
            self._host, self._port, self._path))
        # self._app.run(host=self._host, port=self._port, threaded=True)
        server = Server(self._app)
        server.run()

    def cleanup(self):
        pass

    def add_trigger(self, trigger):
        # This method is called when trigger is created
        pass

    def update_trigger(self, trigger):
        # This method is called when trigger is updated
        pass

    def remove_trigger(self, trigger):
        # This method is called when trigger is deleted
        pass

```
```bash
2018-05-05 07:45:05,862 WARNING [-] Connection to broker lost, trying to re-establish connection...
Traceback (most recent call last):
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/kombu/mixins.py", line 170, in run
    for _ in self.consume(limit=None, **kwargs):
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/kombu/mixins.py", line 186, in consume
    with self.consumer_context(**kwargs) as (conn, channel, consumers):
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/kombu/mixins.py", line 179, in consumer_context
    with self.Consumer() as (connection, channel, consumers):
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/kombu/mixins.py", line 228, in Consumer
    with self._consume_from(*self.get_consumers(cls, channel)) as c:
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/st2common/services/triggerwatcher.py", line 77, in get_consumers
    callbacks=[self.process_task])]
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/kombu/messaging.py", line 386, in __init__
    self.revive(self.channel)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/kombu/messaging.py", line 408, in revive
    self.declare()
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/kombu/messaging.py", line 421, in declare
    queue.declare()
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/kombu/entity.py", line 605, in declare
    self._create_queue(nowait=nowait, channel=channel)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/kombu/entity.py", line 614, in _create_queue
    self.queue_declare(nowait=nowait, passive=False, channel=channel)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/kombu/entity.py", line 649, in queue_declare
    nowait=nowait,
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/amqp/channel.py", line 1151, in queue_declare
    spec.Queue.DeclareOk, returns_tuple=True,
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/amqp/abstract_channel.py", line 79, in wait
    self.connection.drain_events(timeout=timeout)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/amqp/connection.py", line 471, in drain_events
    while not self.blocking_read(timeout):
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/amqp/connection.py", line 477, in blocking_read
    return self.on_inbound_frame(frame)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/amqp/method_framing.py", line 55, in on_frame
    callback(channel, method_sig, buf, None)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/amqp/connection.py", line 481, in on_inbound_method
    method_sig, payload, content,
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/amqp/abstract_channel.py", line 128, in dispatch_method
    listener(*args)
  File "/opt/stackstorm/st2/lib/python2.7/site-packages/amqp/channel.py", line 279, in _on_close
    reply_code, reply_text, (class_id, method_id), ChannelError,
ResourceLocked: Queue.declare: (405) RESOURCE_LOCKED - cannot obtain exclusive access to locked queue 'st2.trigger.watch.sensorwrapper_tutorial_EchoFlaskSensor-a3937549f6' in vhost '/'


```

**Comment by blag** [comment]
@NikosVlagoidis If all you are trying to do is create a bridge to StackStorm with a tiny Flask app, there is a working sensor that implements that - the [GitHubWebhook sensor](https://github.com/StackStorm/GitHubWebhook/blob/master/sensors/webhook_sensor.py#L31).

TL;DR:

```python
class GitHubWebhookSensor(Sensor):
    # ...
    def setup(self):
        # ...
        self.app = Flask(__name__)
        # ...

        @self.app.route('/status')
        def status():
            return json.dumps({"response":"OK"})

        @self.app.route('/webhooks/<path:endpoint>', methods=['POST',])
        def github_events(endpoint):
            # ...

    def run(self):
        self.app.run(host=self.host, port=self.port, debug=False, threaded=True)
```

Just a note: make SURE you run Flask with `debug=False`.

If you are trying to do anything more complicated, I would have the Flask server run standalone and have the Flask server fire off a webhook to StackStorm. We do cover using webhooks in our documentation: https://docs.stackstorm.com/webhooks.html
